### PR TITLE
Support Helm Chart

### LINKED ISSUE #6: Helm chart
Hello,
Do you have plans to create a helm chart for this? I'd like to use for adding e2e test in [KEDA](https://github.com/kedacore/keda) and being honest, having a helm chart it's the easiest way to deploy this :)

### PR DESCRIPTION
**Pull request by everpeace**
depends on #8 
fixes #6 

This PR introduces helm chart support.

- chart itself in `charts/gcp-workload-identity-federation-webhook` dir
- chart is linted in GitHub Actions (`ci.yaml`)
  - chart validation by [chart-testing](https://github.com/helm/chart-testing)
  - generated manifests by default `values.yaml` are validated by [kubeconform](https://github.com/yannh/kubeconform)
- release workflow is also paved by [helm-gh-pages](https://github.com/stefanprodan/helm-gh-pages).
  - I tested the workflow in my forked repo.
  - `index.yaml` and chart tgz will be pushed in `gh-pages` branch. ([sample](https://github.com/everpeace/gcp-workload-identity-federation-webhook/tree/gh-pages) in my fork)
  - So, you can try:
    ```shell
    helm repo add gcp-workload-identity-federation-webhook https://everpeace.github.io/gcp-workload-identity-federation-webhook
    helm repo update
    helm show chart gcp-workload-identity-federation-webhook/gcp-workload-identity-federation-webhook
    ```

## TODO

- [x] prepare empty `gh-pages` branch
- [x] configure Github Pages on `gh-pages` branch

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -82,6 +82,18 @@ docker-push: ## Push docker image with the manager.
...
+
+.PHONY: lint-chart
+lint-chart: helm
+	docker run --rm -v $(PWD):/repo -w /repo \
+		quay.io/helmpack/chart-testing:latest \
+		ct lint --config ct.yaml --print-config
+	# check generated manifest's schema by default values.yaml 
+	for c in $$(ls charts/);  do (cd charts/$${c} && $(HELM) template . | docker run -i --rm -v $(PWD):/repo -w /repo \
+		ghcr.io/yannh/kubeconform:latest \
+		-schema-location default -schema-location 'https://raw.githubusercontent.com/datreeio/CRDs-catalog/main/{{.Group}}/{{.ResourceKind}}_{{.ResourceAPIVersion}}.json' -summary); done
```
Comment (everpeace) [inline Makefile]: linting chart.

File: charts/gcp-workload-identity-federation-webhook/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+# DEVELOPER'S NOTE: version will be mutated in release process
+version: 0.0.0
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+# DEVELOPER'S NOTE: appVersion will be mutated in release process
+appVersion: 0.0.0
```
Comment (everpeace) [inline charts/gcp-workload-identity-federation-webhook/Chart.yaml:25]: `version` and `appVersion` will be injected in `release.yaml`
ref: https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9#discussion_r1029687967

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -20,3 +25,19 @@ jobs:
...
+        script: |
+          const tag = process.env.GITHUB_REF_NAME
+          const no_v = tag.replace('v', '')
+          core.setOutput('no_v', no_v)
+    - name: Publish Helm charts
+      uses: stefanprodan/helm-gh-pages@master
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
+        app_version: ${{steps.tag.outputs.no_v}}
+        chart_version: ${{steps.tag.outputs.no_v}}
```
Comment (everpeace) [inline .github/workflows/release.yaml:43]: This injects `version` and `appVersion` in `Chart.yaml`
ref: https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9#discussion_r1029687424

**Review by everpeace — COMMENTED** [review]
@JorTurFer @superbrothers I wrote a helm chart and ci(lint)/release workflow. could you review this??

**Comment by JorTurFer** [comment]
sure, tomorrow I'll review it. Thanks a lot! 🙇

File: Makefile [hunk 4]
```diff
@@ -136,3 +149,9 @@ $(CONTROLLER_GEN): $(LOCALBIN)
 envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
 $(ENVTEST): $(LOCALBIN)
 	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+
+.PHONY: helm
+HELM_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3"
+helm: $(HELM)
+$(HELM):
+	curl -fsSL $(HELM_INSTALL_SCRIPT) | HELM_INSTALL_DIR=$(LOCALBIN) USE_SUDO=fals bash
```
Comment (superbrothers) [inline Makefile]: Is this typo?

```suggestion
	curl -fsSL $(HELM_INSTALL_SCRIPT) | HELM_INSTALL_DIR=$(LOCALBIN) USE_SUDO=false bash
```

File: Makefile (hunk 1 above)
Comment (superbrothers) [reply Makefile]: nits: I would like to use the container image for helm command. Also, I have tried to organize the Makefile to make it easier to read.

```diff
diff --git a/.github/workflows/ci.yaml b/.github/workflows/ci.yaml
index 716995a..ffd1c1f 100644
--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -28,4 +28,4 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v3
-    - run: mkdir -p /tmp/bin && make lint-chart LOCALBIN=/tmp/bin
+    - run: make lint-chart
diff --git a/Makefile b/Makefile
index 984256c..aa8095f 100644
--- a/Makefile
+++ b/Makefile
@@ -83,16 +83,18 @@ docker-push: ## Push docker image with the manager.
        docker push ${IMG_LATEST}

 ##@ Helm Chart
+DOCKER_RUN ?= docker run --rm -i -v $(PWD):/repo -w /repo -u "$(shell id -u):$(shell id -g)"
+HELM ?= $(DOCKER_RUN) docker.io/alpine/helm:latest
+CT ?= $(DOCKER_RUN) quay.io/helmpack/chart-testing:latest ct
+KUBECONFORM ?= $(DOCKER_RUN) ghcr.io/yannh/kubeconform:latest

 .PHONY: lint-chart
-lint-chart: helm
-       docker run --rm -v $(PWD):/repo -w /repo \
-               quay.io/helmpack/chart-testing:latest \
-               ct lint --config ct.yaml --print-config
-       # check generated manifest's schema by default values.yaml
-       for c in $$(ls charts/);  do (cd charts/$${c} && $(HELM) template . | docker run -i --rm -v $(PWD):/repo -w /repo \
-               ghcr.io/yannh/kubeconform:latest \
-               -schema-location default -schema-location 'https://raw.githubusercontent.com/datreeio/CRDs-catalog/main/{{.Group}}/{{.ResourceKind}}_{{.ResourceAPIVersion}}.json' -summary); done
+lint-chart:
+       $(CT) lint --config ct.yaml --print-config
+       # check generated manifest's schema by default values.yaml
+       for chart in $$(ls charts/); do \
+               $(HELM) template "charts/$${chart}" | $(KUBECONFORM) -schema-location default -schema-location 'https://raw.githubusercontent.com/datreeio/CRDs-catalog/main/{{.Group}}/{{.ResourceKind}}_{{.ResourceAPIVersion}}.json' -summary; \
+       done

 ##@ Deployment

@@ -128,7 +130,6 @@ $(LOCALBIN):
 KUSTOMIZE ?= $(LOCALBIN)/kustomize
 CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
 ENVTEST ?= $(LOCALBIN)/setup-envtest
-HELM ?= $(LOCALBIN)/helm

 ## Tool Versions
 KUSTOMIZE_VERSION ?= v4.5.5
@@ -149,9 +150,3 @@ $(CONTROLLER_GEN): $(LOCALBIN)
 envtest: $(ENVTEST) ## Download envtest-setup locally if necessary.
 $(ENVTEST): $(LOCALBIN)
        GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
-
-.PHONY: helm
-HELM_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3"
-helm: $(HELM)
-$(HELM):
-       curl -fsSL $(HELM_INSTALL_SCRIPT) | HELM_INSTALL_DIR=$(LOCALBIN) USE_SUDO=fals bash
```

File: README.md [hunk 5]
```diff
@@ -168,10 +168,21 @@ Usage of /gcp-workload-identity-federation-webhook:
 ### Pre-requisites
 
 - cert-manager: See [cert-manager installation](https://cert-manager.io/docs/installation/)
-- (optional) prometheus-operator: See https://github.com/prometheus-operator/prometheus-operator
+- prometheus-operator: See https://github.com/prometheus-operator/prometheus-operator
```
Comment (superbrothers) [inline README.md]: Without Prometheus Operator, the installation of this controller will fail. However, Prometheus Operator is not required for this controller to work, so optional is preferred.

```diff
diff --git a/README.md b/README.md
index 39a3e9a..f0a4d30 100644
--- a/README.md
+++ b/README.md
@@ -168,7 +168,7 @@ Usage of /gcp-workload-identity-federation-webhook:
 ### Pre-requisites

 - cert-manager: See [cert-manager installation](https://cert-manager.io/docs/installation/)
-- prometheus-operator: See https://github.com/prometheus-operator/prometheus-operator
+- (optional) prometheus-operator: See https://github.com/prometheus-operator/prometheus-operator

 ### Deploy

diff --git a/charts/gcp-workload-identity-federation-webhook/templates/controller-manager-metrics-monitor.yaml b/charts/gcp-workload-identity-federation-webhook/templates/controller-manager-metrics-monitor.yaml
index 001654a..d9917eb 100644
--- a/charts/gcp-workload-identity-federation-webhook/templates/controller-manager-metrics-monitor.yaml
+++ b/charts/gcp-workload-identity-federation-webhook/templates/controller-manager-metrics-monitor.yaml
@@ -1,3 +1,4 @@
+{{- if .Values.serviceMonitor.enabled }}
 apiVersion: monitoring.coreos.com/v1
 kind: ServiceMonitor
 metadata:
@@ -17,3 +18,4 @@ spec:
     matchLabels:
       control-plane: controller-manager
     {{- include "gcp-workload-identity-federation-webhook.selectorLabels" . | nindent 6 }}
+{{- end }}
diff --git a/charts/gcp-workload-identity-federation-webhook/values.yaml b/charts/gcp-workload-identity-federation-webhook/values.yaml
index c09c3c8..1740ada 100644
--- a/charts/gcp-workload-identity-federation-webhook/values.yaml
+++ b/charts/gcp-workload-identity-federation-webhook/values.yaml
@@ -49,6 +49,9 @@ controllerManager:

 kubernetesClusterDomain: cluster.local

+serviceMonitor:
+  enabled: false
+
 metricsService:
   ports:
   - name: metrics
```

File: README.md [hunk 6]
```diff
@@ -168,10 +168,21 @@ Usage of /gcp-workload-identity-federation-webhook:
...
 
 ### Deploy
 
+#### Helm chart
+
+```shell
+$ helm repo add gcp-workload-identity-federation-webhook https://pfnet-research.github.io/gcp-workload-identity-federation-webhook
+$ helm repo update
+$ helm install gcp-wif-webhook gcp-workload-identity-federation-webhook/gcp-workload-identity-federation-webhook \
+    --namespace gcp-wif-webhook-system　--create-namespace
```
Comment (superbrothers) [inline README.md]:
```suggestion
    --namespace gcp-wif-webhook-system --create-namespace
```

File: charts/gcp-workload-identity-federation-webhook/templates/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,106 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ include "gcp-workload-identity-federation-webhook.fullname" . }}-controller-manager
```
Comment (superbrothers) [inline charts/gcp-workload-identity-federation-webhook/templates/deployment.yaml:4]: metadata.namespace field is not automatically set if using `helm template` command.

```
$ helm version
version.BuildInfo{Version:"v3.9.0", GitCommit:"7ceeda6c585217a19a1131663d8cd1f7d641b2a7", GitTreeState:"clean", GoVersion:"go1.17.5"}
$ helm template charts/gcp-workload-identity-federation-webhook | grep metadata -A 10 -B 10 | head -n 15
---
# Source: gcp-workload-identity-federation-webhook/templates/deployment.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: release-name-gcp-wif-webhook-controller-manager
  labels:
    helm.sh/chart: gcp-workload-identity-federation-webhook-0.0.0
    app.kubernetes.io/name: gcp-wif-webhook
    app.kubernetes.io/instance: release-name
    app.kubernetes.io/version: "0.0.0"
    app.kubernetes.io/managed-by: Helm
---
# Source: gcp-workload-identity-federation-webhook/templates/manager-rbac.yaml
apiVersion: rbac.authorization.k8s.io/v1
```

The below change is user-friendly for `helm template` users.

```suggestion
metadata:
  name: {{ include "gcp-workload-identity-federation-webhook.fullname" . }}-controller-manager
  namespace: {{ .Release.Namespace }}
```

File: Makefile (hunk 1 above)
Comment (everpeace) [reply Makefile]: ah, that makes sense. thanks.
https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9/commits/2c14df5c2f7633505881fdedf553fa1e56abef2b

File: Makefile (hunk 4 above)
Comment (everpeace) [reply Makefile]: This won't be needed as we use docker for helm command in Makefile.
ref: https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9#discussion_r1030981642

File: README.md (hunk 5 above)
Comment (everpeace) [reply README.md]: 👍 
https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9/commits/99f27b698705b9ef229633ebef1f9e9027a2a605

File: charts/gcp-workload-identity-federation-webhook/templates/deployment.yaml (hunk 7 above)
Comment (everpeace) [reply charts/gcp-workload-identity-federation-webhook/templates/deployment.yaml:4]: 👍 I put `metadata.name` to all the namespaced resources.
https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9/commits/2ab0216a9e77c32b6590571d798f0a91d3c5cc30

File: README.md (hunk 6 above)
Comment (everpeace) [reply README.md]: thanks.
https://github.com/pfnet-research/gcp-workload-identity-federation-webhook/pull/9/commits/178584013f78872e7eb87362bdda529a1fd3cd66

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -23,3 +23,9 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - run: make docker-build
+
+  lint-chart:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - run: mkdir -p /tmp/bin && make lint-chart LOCALBIN=/tmp/bin
```
Comment (superbrothers) [inline .github/workflows/ci.yaml]: With 2c14df5, `LOCALBIN` environmental variable is no longer needed.

```suggestion
    - make lint-chart
```

File: Makefile [hunk 9]
```diff
@@ -116,6 +131,7 @@ $(LOCALBIN):
 KUSTOMIZE ?= $(LOCALBIN)/kustomize
 CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
 ENVTEST ?= $(LOCALBIN)/setup-envtest
+HELM ?= $(LOCALBIN)/helm
```
Comment (superbrothers) [inline Makefile]: With 2c14df5, this line is no longer needed.

```suggestion
```

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -23,3 +23,9 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - run: make docker-build
+
+  lint-chart:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - make lint-chart
```
Comment (superbrothers) [inline .github/workflows/ci.yaml]: I'm sorry. I wrote it wrong.

```suggestion
    - run: make lint-chart
```

**Review by superbrothers — DISMISSED** [review]

File: .github/workflows/ci.yaml (hunk 10 above)
Comment (everpeace) [reply .github/workflows/ci.yaml]: sorry, too🙇‍♂️

**Review by superbrothers — APPROVED** [review]

**Review by JorTurFer — APPROVED** [review]
LGTM! I'll add it to KEDA e2e tests once it's available and I can provide more info about if it works :) 
Awesome job 🙇 
Just one tip working with helm charts, you can use Kind in the GH Runner to deploy your charts to a local cluster, that could give you more feedback about invalid APIs or sections in the manifests. For instance, in KEDA we use it to validate the chart with [several k8s versions](https://github.com/kedacore/charts/blob/main/.github/workflows/ci-core.yml#L32-L127)

**Comment by everpeace** [comment]
> Just one tip working with helm charts, you can use Kind in the GH Runner to deploy your charts to a local cluster, that could give you more feedback about invalid APIs or sections in the manifests. For instance, in KEDA we use it to validate the chart with [several k8s versions](https://github.com/kedacore/charts/blob/main/.github/workflows/ci-core.yml#L32-L127)

Cool! I will try to do this in another PR. 
I'm merging this PR for now and will release `v0.1.1` to make helm chart available.

**Pull request merged**